### ISSUE TITLE
Bug: unexpected automatic dtype determination of `gvar.BufferDict`

### ISSUE THREAD

**Issue by Gattocrucco** [gplepage/gvar#35]
On creation, if a data type is not specified, `gvar.BufferDict` upcasts to int64 or float64:

```python
import gvar
import numpy as np
print(gvar.BufferDict(a=np.ones(1, bool)).dtype)
print(gvar.BufferDict(a=np.ones(1, 'int32')).dtype)
print(gvar.BufferDict(a=np.ones(1, 'uint64')).dtype)
print(gvar.BufferDict(a=np.ones(1, 'float32')).dtype)
```

```
int64
int64
float64
float64
```

This is unexpected, in particular when I pass explicitly typed numpy arrays as input. Something like this might make sense as a policy if the dtype of a BufferDict was immutable, to have a very simple rule for its automatic determination, however it is mutable:

```python
bd = gvar.BufferDict(a=1)
print(bd.dtype)
bd['b'] = 2.3
print(bd.dtype)
```

```
int64
float64
```

I think that the automatic data type determination should work as in numpy.

### version
gvar 11.11.1

**Comment by gplepage** [comment]
This is intentional. 

People often mix types in a ``BufferDict`` -- for example, integers with floats (eg, ``[1.2,3,4.6]``), or floats with ``GVar``s. I don't want the ``BufferDict`` type determined by the first entry in the dictionary; that seems too subtle and error prone. Instead ``BufferDict`` tries to find a data type that can preserve the information in all of its entries (eg, converting integers to floats rather than the other way around). This means the underlying data type has  to be able to change when new items are added. 

``numpy`` does the same thing when you create an array from a list of mixed types or concatenating arrays of different types. ``BufferDict`` uses ``numpy``'s algorithm for determining the overall type.

**Comment by Gattocrucco** [comment]
I agree that `BufferDict` should try to determine an overall datatype the way you describe, but I disagree that the current behavior matches numpy's algorithm. `BufferDict` seems to "overreact" and upcast the data type even when it is not necessary. Example:

```python
import gvar
import numpy as np

print(np.array([1, 1.]).dtype) # -> float64
print(gvar.BufferDict(a=1, b=1.).dtype) # -> float64
print(np.array(True).dtype) # -> bool
print(gvar.BufferDict(a=True).dtype) # -> int64  <=== not ok
print(np.array(1, 'int32').dtype) # -> int32
print(gvar.BufferDict(a=np.array(1, 'int32')).dtype) # -> int64 <=== not ok
```

**Comment by gplepage** [comment]
I agree that these last examples aren't right. The problem will be fixed in the next release.

**Comment by gplepage** [comment]
This problem is fixed in v11.11.2.

**Comment by Gattocrucco** [comment]
Thanks.
